### ISSUE TITLE
Crash when performing an agent-check in 2.4.1, worked in 2.3.10

### ISSUE THREAD

**Issue by donothangbagsfromthehandlebar** [haproxy/haproxy#1302]
## Detailed description of the problem

haproxy 2.4.1 seems to crash when performing an `agent-check` against a backend server

## Expected behavior

No crash as in haproxy 2.3.10

## Steps to reproduce the behavior

1. On the backend server, enable a haproxy agent protocol implementation. Sample output:
```
$ socat tcp4:srv01.local:9081 -
100% maxconn:2000 ready
```
2. In haproxy, configure backend servers to perform `agent-check` as well as http check
3. Start haproxy using `systemctl start haproxy.service`
4. Observe the haproxy logs

Result: Frequent crashes and restarts, potentially due to the agent-check. Disabling the agent-check results in a working setup.

## Do you have any idea what may have caused this?

Upgrade to haproxy 2.4.1, potentially together with a slightly incorrect agent protocol implementation

## Do you have an idea how to solve the issue?

For now, continue to use haproxy 2.3.10

## What is your configuration?

```
global
   log /dev/log local0
   log /dev/log local1 notice
   chroot /var/lib/haproxy
   stats timeout 30s
   user haproxy
   group haproxy
   daemon

   stats socket /var/run/haproxy.sock mode 600 level admin

defaults
   log global
   mode http
   option httplog
   option dontlognull
   option redispatch
   option prefer-last-server
   option http-keep-alive
   timeout connect 300ms
   retries 1
   timeout client 30m
   timeout server 30m
   timeout tunnel 31m


cache cache01
  total-max-size 100
  max-age 10485760
  max-object-size 10485760

backend servers

   http-request cache-use cache01
   http-response cache-store cache01

   # For debug purposes
   http-response set-header X-Server %s
   http-response set-header X-Cache-Status HIT if { res.cache_hit }
   http-response set-header X-Cache-Status MISS if !{ res.cache_hit }

   option httpchk GET /status HTTP/1.0

   server srv01 srv01.local:8080 weight 100 check agent-check agent-port 9081
   server srv02 srv02.local:8080 weight 100 check agent-check agent-port 9081

```

## Output of `haproxy -vv` and `uname -a`

```
# ./haproxy -vv
HAProxy version 2.4.1-1ce7d49 2021/06/17 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.1.html
Running on: Linux 4.18.0-240.22.1.el8_3.x86_64 #1 SMP Thu Apr 8 19:01:30 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1g FIPS  21 Apr 2020
Running on OpenSSL version : OpenSSL 1.1.1g FIPS  21 Apr 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.4.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.42 2018-03-20
Running on PCRE version : 8.42 2018-03-20
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.4.1 20200928 (Red Hat 8.4.1-1)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace

```

```
# uname -a
Linux haproxy01.local 4.18.0-240.22.1.el8_3.x86_64 #1 SMP Thu Apr 8 19:01:30 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

## If HAProxy crashed: Last outputs and backtraces

journalctl:
```
Jun 23 01:33:43 haproxy01.local systemd[1]: haproxy.service: Service RestartSec=100ms expired, scheduling restart.
Jun 23 01:33:43 haproxy01.local systemd[1]: haproxy.service: Scheduled restart job, restart counter is at 100.
Jun 23 01:33:43 haproxy01.local systemd[1]: Stopped HAProxy Load Balancer.
Jun 23 01:33:43 haproxy01.local systemd[1]: Starting HAProxy Load Balancer...
Jun 23 01:33:43 haproxy01.local haproxy[493285]: [NOTICE]   (493285) : New worker #1 (493287) forked
Jun 23 01:33:43 haproxy01.local systemd[1]: Started HAProxy Load Balancer.
Jun 23 01:33:47 haproxy01.local haproxy[493285]: Thread 1 is about to kill the process.
Jun 23 01:33:47 haproxy01.local haproxy[493285]: *>Thread 1 : id=0x7f654f648080 act=1 glob=1 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              stuck=1 prof=0 harmless=0 wantrdv=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              cpu_ns: poll=3863874 now=3003225720 diff=2999361846
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              curr_task=0x1159bc0 (tasklet) calls=2
Jun 23 01:33:47 haproxy01.local haproxy[493285]:                fct=0x545790(event_srv_chk_io) ctx=0x1109820
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              call trace(21):
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x5ce9df [48 83 c4 10 5b 5d 41 5c]: wdt_handler+0xff/0x10c
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              | 0x7f654e9d8b20 [48 c7 c0 0f 00 00 00 0f]: libpthread:+0x12b20
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x4daf17 [48 85 ca 75 7c 48 89 f2]: main+0xb7307
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x4b114a [48 8b 74 24 08 48 85 f6]: tcpcheck_agent_expect_reply+0x1ba/0x99c
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x4b5fda [41 89 c6 8b 05 85 cd 3f]: tcpcheck_eval_expect+0x11a/0x795
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x4b6a86 [85 c0 0f 85 d2 01 00 00]: tcpcheck_main+0x396/0xc00
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x5453bc [48 8b 6b 38 41 89 c5 48]: main+0x1217ac
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x54579d [31 c0 48 83 c4 08 c3 66]: event_srv_chk_io+0xd/0x14
Jun 23 01:33:47 haproxy01.local haproxy[493285]:   Thread 2 : id=0x7f654ba45700 act=0 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              stuck=0 prof=0 harmless=1 wantrdv=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              cpu_ns: poll=1898360 now=2742393 diff=844033
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              curr_task=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:   Thread 3 : id=0x7f6543fff700 act=1 glob=1 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              stuck=0 prof=0 harmless=0 wantrdv=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              cpu_ns: poll=2040790 now=1687714503 diff=1685673713
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              curr_task=0x115ae40 (task) calls=3 last=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:                fct=0x547150(process_chk) ctx=0x11095b8
Jun 23 01:33:47 haproxy01.local haproxy[493285]:   Thread 4 : id=0x7f65437fe700 act=0 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              stuck=0 prof=0 harmless=1 wantrdv=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              cpu_ns: poll=2081187 now=2991182 diff=909995
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              curr_task=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:   Thread 5 : id=0x7f6542ffd700 act=1 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              stuck=0 prof=0 harmless=0 wantrdv=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              cpu_ns: poll=2508545 now=709871385 diff=707362840
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              curr_task=0x11656d0 (task) calls=3 last=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:                fct=0x547150(process_chk) ctx=0x110a8d8
Jun 23 01:33:47 haproxy01.local haproxy[493285]:   Thread 6 : id=0x7f65427fc700 act=0 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              stuck=0 prof=0 harmless=1 wantrdv=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              cpu_ns: poll=1519582 now=2490978 diff=971396
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              curr_task=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:   Thread 7 : id=0x7f6541ffb700 act=0 glob=0 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              stuck=0 prof=0 harmless=1 wantrdv=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              cpu_ns: poll=2111774 now=3086831 diff=975057
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              curr_task=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:  >Thread 8 : id=0x7f65417fa700 act=1 glob=1 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              stuck=1 prof=0 harmless=0 wantrdv=0
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              cpu_ns: poll=1539165 now=2010947175 diff=2009408010
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              curr_task=0x1159c10 (tasklet) calls=2
Jun 23 01:33:47 haproxy01.local haproxy[493285]:                fct=0x545790(event_srv_chk_io) ctx=0x110ab40
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              call trace(16):
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x5743ac [85 c0 75 18 48 81 c4 c8]: ha_dump_backtrace+0x2c/0x2b8
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x574c6f [48 83 c4 38 5b 5d 41 5c]: ha_thread_dump+0x22f/0x281
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x574d36 [48 8b 05 1b da 3c 00 48]: debug_handler+0x66/0x10e
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              | 0x7f654e9d8b20 [48 c7 c0 0f 00 00 00 0f]: libpthread:+0x12b20
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x4daf17 [48 85 ca 75 7c 48 89 f2]: main+0xb7307
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x4b114a [48 8b 74 24 08 48 85 f6]: tcpcheck_agent_expect_reply+0x1ba/0x99c
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x4b5fda [41 89 c6 8b 05 85 cd 3f]: tcpcheck_eval_expect+0x11a/0x795
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x4b6a86 [85 c0 0f 85 d2 01 00 00]: tcpcheck_main+0x396/0xc00
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x5453bc [48 8b 6b 38 41 89 c5 48]: main+0x1217ac
Jun 23 01:33:47 haproxy01.local haproxy[493285]:              |       0x54579d [31 c0 48 83 c4 08 c3 66]: event_srv_chk_io+0xd/0x14
Jun 23 01:33:47 haproxy01.local haproxy[493285]: [NOTICE]   (493285) : haproxy version is 2.4.1-1ce7d49
Jun 23 01:33:47 haproxy01.local haproxy[493285]: [NOTICE]   (493285) : path to executable is /usr/local/sbin/haproxy
Jun 23 01:33:47 haproxy01.local haproxy[493285]: [ALERT]    (493285) : Current worker #1 (493287) exited with code 134 (Aborted)
Jun 23 01:33:47 haproxy01.local haproxy[493285]: [ALERT]    (493285) : exit-on-failure: killing every processes with SIGTERM
Jun 23 01:33:47 haproxy01.local haproxy[493285]: [WARNING]  (493285) : All workers exited. Exiting... (134)
```

## Additional information (if helpful)

Both haproxy 2.4.1 and 2.3.10 are compiled from sources on the same machine as it is running

haproxy 2.3.10 works with the same agent implementation and haproxy configuration
```
# haproxy -vv
HA-Proxy version 2.3.10-4764f0e 2021/04/23 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2022.
Known bugs: http://www.haproxy.org/bugs/bugs-2.3.10.html
Running on: Linux 4.18.0-240.22.1.el8_3.x86_64 #1 SMP Thu Apr 8 19:01:30 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1g FIPS  21 Apr 2020
Running on OpenSSL version : OpenSSL 1.1.1g FIPS  21 Apr 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.4.3
Built with network namespace support.
Built with the Prometheus exporter as a service
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.42 2018-03-20
Running on PCRE version : 8.42 2018-03-20
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.3.1 20191121 (Red Hat 8.3.1-5)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace

```

**Comment by a-denoyelle** [comment]
Thanks for your report. This issue has already been identified internally. It is caused by a deadlock when changing maxconn during agent-check. A fix is already present in the development tree.

commit 0274286dd3768c0d5e58588a1cb7e7e710fbc9d4
BUG/MAJOR: server: fix deadlock when changing maxconn via agent-check

**Comment by capflam** [comment]
The above fix was backported in 2.4 and 2.3. I'm closing this issue now.
